module/vision/BallDetector/src/BallDetector.cpp

```diff
@@ -117,126 +118,155 @@ namespace module::vision {
+                // currently visible
                 if (clusters.empty()) {
                     log<NUClear::DEBUG>("Found no balls.");
                     emit(std::move(balls));
                     return;
                 }
 
+                // Reserve the memory prematurely for efficiency
                 balls->balls.reserve(clusters.size());
 
-                balls->id        = horizon.id;
-                balls->timestamp = horizon.timestamp;
-                balls->Hcw       = horizon.Hcw;
+                balls->id        = horizon.id;         // subsumption id
+                balls->timestamp = horizon.timestamp;  // time when the image was taken
+                balls->Hcw       = horizon.Hcw;        // world to camera transform at the time the image was taken
 
+                // Check each cluster to see if it's a valid ball
                 for (auto& cluster : clusters) {
                     Ball b;
 
                     // Average the cluster to get the cones axis
                     Eigen::Vector3f axis = Eigen::Vector3f::Zero();
+                    // Add up all the unit vectors of each point (camera to point in world space) in the cluster
                     for (const auto& idx : cluster) {
                         axis += rays.col(idx);
                     }
-                    axis /= cluster.size();
+                    axis /= cluster.size();  // can this line be skipped if it's being normalised in the next line?
                     axis.normalize();
 
                     // Find the ray with the greatest distance from the axis
-                    // Should we use the average distance instead?
                     float radius = 1.0f;
                     for (const auto& idx : cluster) {
                         const Eigen::Vector3f& ray(rays.col(idx));
-                        if (axis.dot(ray) < radius) {
-                            radius = axis.dot(ray);
-                        }
+                        // arccos of the dot product gives the angle between the rays
+                        // the smaller arccos is, the bigger the angle - if between 0 and pi
+                        radius = axis.dot(ray) < radius ? axis.dot(ray) : radius;
                     }
 
-                    // Ball cam space info
-                    b.cone.axis   = horizon.Hcw.topLeftCorner<3, 3>().cast<float>() * axis;
-                    b.cone.radius = radius;
+                    // Set cone information for the ball
+                    // The rays are in world space, multiply by Rcw to get the axis in camera space
+                    b.cone.axis = horizon.Hcw.topLeftCorner<3, 3>().cast<float>()
+                                  * axis;    // Hcw is not an affine transform type (cannot use linear())
+                    b.cone.radius = radius;  // arccos(radius) is the angle between the furthest vectors
```

Comment:
Just make it an Affine3d if you need Affine3d

`Eigen::Affine3d(horizon.Hcw).cast<float>().rotation() * axis`

Reply:
Hmmm I guess this is easier to read